Split out IO logic from high level functions (#392)

jstockwin opened the pull request:
**Pull request**

Adds two new functions (`extract_text_from_io` and `extract_pages_from_io`) to `high_level.py`, allowing you to call `extract_text` and `extract_pages` without having the file locally.

Closes #392 

**Notes**
- I've created the new functions, `*_from_io` to avoid breaking changes to the original functions. This is not very consistent with the `extract_text_to_fp` at the top of the file, perhaps. Happy to implement alternatives.
- I've used `inf` as the new IO parameter, even though I don't really like the name. This is to be consistent with `extract_text_to_fp`. Other places in the code use `fp`. Happy to change.
- I added the test in a new file, as the existing code has `test_highlevel_extracttext.py`. Is this correct, or should this all be in one `test_high_level.py`?


**How Has This Been Tested?**

I've added tests for `extract_text_from_io`, but couldn't find any existing tests for `extract_pages`?

**Checklist**

- [X] I have added tests that prove my fix is effective or that my feature 
  works
- [X] I have added docstrings to newly created methods and classes
- [X] I have optimized the code at least one time after creating the initial 
  version
- [X] I have updated the [README.md](../README.md) or I am verified that this
  is not necessary (not necessary)
- [X] I have updated the [readthedocs](../docs/source) documentation or I 
  verified that this is not necessary (not necessary as the read the docs page on this is auto generated from docstrings)
- [X] I have added a consice human-readable description of the change to 
  [CHANGELOG.md](../CHANGELOG.md)

jstockwin commented:
@pietermarsman PR updated to keep the same function name and simply allow either a string or a file-like object. I've added an `open_filename` context manager to help do this. I was tempted to put this in `utils`, but it didn't quite seem to fit...

I've essentially followed advice from here about how to do this... https://stackoverflow.com/questions/7268353/how-to-accept-both-filenames-and-file-like-objects-in-python-functions

pietermarsman commented on CHANGELOG.md:
```diff
@@ -8,6 +8,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Grouping of text lines outside of parent container bounding box ([#386](https://github.com/pdfminer/pdfminer.six/pull/386))
 
+#### Changed
+- Added `extract_text_from_io` and `extract_pages_from_io` high level functions ([#392](https://github.com/pdfminer/pdfminer.six/pull/392))
```
Outdated

pietermarsman commented on pdfminer/high_level.py:
```diff
@@ -13,6 +13,28 @@
 from .pdfpage import PDFPage
 
 
+class open_filename(object):
```
I agree that this should be moved to utils.py. If the current content of utils.py bothers you, perhaps we can move some of it to a better place in another issue/pr.

jstockwin commented:
@pietermarsman thanks. I've updated the first commit to resolve contributing being out of date, and added a new commit moving the `open_filename` to `utils`.

Please take another look when you get chance.
